skrifa/src/outline/glyf/hint/engine/definition.rs

```diff
@@ -0,0 +1,491 @@
+    /// and <https://gitlab.freedesktop.org/freetype/freetype/-/blob/57617782464411201ce7bbc93b086c1b4d7d84a5/src/truetype/ttinterp.c#L3623>
+    pub(super) fn op_call(&mut self) -> OpResult {
+        let f = self.value_stack.pop()?;
+        self.do_call(DefKind::Function, 1, f)
+    }
+
+    /// Loop and call function.
+    ///
+    /// LOOPCALL[] (0x2a)
+    ///
+    /// Pops: f: function identifier number
+    ///       count: number of times to call the function
+    ///
+    /// Calls the function f, count number of times.
+    ///
+    /// See <https://learn.microsoft.com/en-us/typography/opentype/spec/tt_instructions#loop-and-call-function>
+    /// and <https://gitlab.freedesktop.org/freetype/freetype/-/blob/57617782464411201ce7bbc93b086c1b4d7d84a5/src/truetype/ttinterp.c#L3704>
+    pub(super) fn op_loopcall(&mut self) -> OpResult {
+        let f = self.value_stack.pop()?;
+        let count = self.value_stack.pop()?;
+        self.loop_budget.doing_loop_call(count as usize)?;
+        self.do_call(DefKind::Function, count as u32, f)
+    }
+
+    /// Instruction definition.
+    ///
+    /// IDEF[] (0x89)
+    ///
+    /// Pops: opcode
+    ///
+    /// Begins the definition of an instruction. The instruction definition
+    /// terminates when at ENDF, which is encountered in the instruction
+    /// stream. Subsequent executions of the opcode popped will be directed
+    /// to the contents of this instruction definition (IDEF). IDEFs must be
+    /// defined in the Font Program or the CVT Program; attempts to invoke the
+    /// IDEF instruction within a glyph program will result in an error. An
+    /// IDEF affects only undefined opcodes. If the opcode in question is
+    /// already defined, the interpreter will ignore the IDEF. This is to be
+    /// used as a patching mechanism for future instructions. Instructions
+    /// may not exceed 64K in size.
+    ///
+    /// See <https://learn.microsoft.com/en-us/typography/opentype/spec/tt_instructions#instruction-definition>
+    /// and <https://gitlab.freedesktop.org/freetype/freetype/-/blob/57617782464411201ce7bbc93b086c1b4d7d84a5/src/truetype/ttinterp.c#L3788>
+    pub(super) fn op_idef(&mut self) -> OpResult {
+        let opcode = self.value_stack.pop()?;
+        self.do_def(DefKind::Instruction, opcode)
+    }
+
+    /// Catch all for unhandled opcodes which will attempt to dispatch to a
+    /// user defined instruction.
+    pub(super) fn op_unknown(&mut self, opcode: u8) -> OpResult {
+        match self.do_call(DefKind::Instruction, 1, opcode as i32) {
+            // Remap an invalid definition error to unhandled opcode
+            Err(HintErrorKind::InvalidDefinition(opcode)) => Err(HintErrorKind::UnhandledOpcode(
+                Opcode::from_byte(opcode as u8),
+            )),
+            result => result,
+        }
+    }
+
+    /// Common code for FDEF and IDEF.
+    fn do_def(&mut self, kind: DefKind, key: i32) -> OpResult {
```

Comment:
several things that note they "may not exceed 64K in size" call this but I'm not immediately spotting where we enforce that. If we do, perhaps a comment?

I also don't immediately see a test for an excessively long function or instructin.

Reply:
Good catch. We don’t enforce that limit because FT doesn’t. I’m fine with adding a check if you’d prefer but also don’t see the harm in leaving it out.